### PR TITLE
Vision sim 🤯 

### PR DESCRIPTION
**Pull request by Ishan1522**

### CODE REVIEW

**Comment by Ishan1522** [comment]
plz review dis gonna be very cool

**Comment by Ishan1522** [comment]
the # of files changed will decrease once #63 is merged

**Comment by Ishan1522** [comment]
i cooked guys

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -271,48 +270,59 @@ private void configureButtonBindings() {
...
     // // // Reset robot odometry based on vision pose measurement from april tags
-    // driverLeftDirectionPad.onTrue(new InstantCommand(() ->
-    // driveSubsystem.resetOdometry(visionSubsystem.getLastSeenPose())));
-    // // driverLeftDpad.onTrue(new InstantCommand(() -> driveSubsystem.resetOdometry(new
+    driverLeftDirectionPad.onTrue(
+        new InstantCommand(() -> swerveDrive.resetPosition(visionSubsystem.getLastSeenPose())));
+    // // driverLeftDpad.onTrue(new InstantCommand(() -> swerveDrive.resetOdometry(new
     // Pose2d(15.251774787902832, 5.573054313659668, Rotation2d.fromRadians(3.14159265)))));
-    // // driverBButton.whileTrue(new ShootPass(driveSubsystem, shooterSubsystem, pivotSubsystem,
+    // // driverBButton.whileTrue(new ShootPass(swerveDrive, shooterSubsystem, pivotSubsystem,
```
Comment (JacksonElia) [inline src/main/java/frc/robot/RobotContainer.java:305]: we got sooooo many comments

File: src/main/java/frc/robot/commands/drive/DriveCommandBase.java [hunk 2]
```diff
@@ -39,43 +42,43 @@ public void execute() {
     swerveDrive.addPoseEstimatorSwerveMeasurement();
     vision.setHeadingInfo(
         swerveDrive.getPose().getRotation().getDegrees(), swerveDrive.getGyroRate());
-    calculatePoseFromLimelight(VisionConstants.SHOOTER_LIMELIGHT_NUMBER);
-    calculatePoseFromLimelight(VisionConstants.FRONT_LEFT_LIMELIGHT_NUMBER);
-    calculatePoseFromLimelight(VisionConstants.FRONT_RIGHT_LIMELIGHT_NUMBER);
+    calculatePoseFromLimelight(Limelight.SHOOTER);
```
Comment (JacksonElia) [inline src/main/java/frc/robot/commands/drive/DriveCommandBase.java:45]: what is Limelight.SHOOTER? Is Limelight the constants class now?

Comment (Ishan1522) [reply src/main/java/frc/robot/commands/drive/DriveCommandBase.java:45]: its an enum within visionconstants

Comment (JacksonElia) [reply src/main/java/frc/robot/commands/drive/DriveCommandBase.java:43]: nice, this is better

Comment (Ishan1522) [reply src/main/java/frc/robot/commands/drive/DriveCommandBase.java:43]: Bro is not cooking with these comments

File: src/main/java/frc/robot/subsystems/swerve/odometryThread/OdometryThread.java [hunk 3]
```diff
@@ -47,7 +48,7 @@ static Queue<Angle> registerInput(Supplier<Angle> supplier) {
   }
 
   static OdometryThread createInstance(DeviceCANBus canBus) {
-    return switch (Robot.CURRENT_ROBOT_MODE) {
+    return switch (Constants.currentMode) {
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/swerve/odometryThread/OdometryThread.java:51]: What is this? Why do we just have a constants class with stuff inside it and why is current mode not SCREAMING_SNAKE_CASE?

File: src/main/java/frc/robot/subsystems/vision/SimulatedVision.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+import org.photonvision.simulation.VisionSystemSim;
+import org.photonvision.targeting.PhotonPipelineResult;
+
+// Simulate the vision system.
+// Please see the following link for example code
+// https://github.com/PhotonVision/photonvision/blob/2a6fa1b6ac81f239c59d724da5339f608897c510/photonlib-java-examples/swervedriveposeestsim/src/main/java/frc/robot/Vision.java
+public class SimulatedVision extends PhysicalVision {
+  // private final PhotonCamera shooterCamera = new PhotonCamera("shooterCamera");
+  // private final PhotonCamera frontLeftCamera = new PhotonCamera("frontLeftCamera");
+  // private final PhotonCamera frontRightCamera = new PhotonCamera("frontRightCamera");
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/vision/SimulatedVision.java:22]: Delete unecessary comments

File: src/main/java/frc/robot/subsystems/vision/SimulatedVision.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    cameraProperties.setAvgLatencyMs(20);
+    cameraProperties.setLatencyStdDevMs(5);
+
+    // Create a PhotonCameraSim which will update the linked PhotonCamera's values
+    // with visible
+    // targets.
+    // Instance variables
+    shooterCameraSim =
+        new PhotonCameraSim(
+            Limelight.SHOOTER.getSimulationCamera(Limelight.SHOOTER), cameraProperties);
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/vision/SimulatedVision.java:61]: I feel like this syntax could be cleaned up. If the method is coming from the shooter limelight, we shouldn't have to pass it in

File: src/main/java/frc/robot/subsystems/vision/SimulatedVision.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    frontLeftCameraSim =
+        new PhotonCameraSim(
+            Limelight.FRONT_LEFT.getSimulationCamera(Limelight.FRONT_LEFT), cameraProperties);
+    frontRightCameraSim =
+        new PhotonCameraSim(
+            Limelight.FRONT_RIGHT.getSimulationCamera(Limelight.FRONT_RIGHT), cameraProperties);
+
+    Transform3d robotToShooterCamera =
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/vision/SimulatedVision.java:69]: All of these Transform3ds should be constants.

File: src/main/java/frc/robot/subsystems/vision/SimulatedVision.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+  public void updateInputs(VisionInputs inputs) {
+    // Abuse the updateInputs periodic call to update the sim
+
+    // Move the vision sim robot on the field
+    if (robotSimulationPose.get() != null) {
+      visionSim.update(robotSimulationPose.get());
+      Logger.recordOutput("Vision/SimIO/updateSimPose", robotSimulationPose.get());
+    }
+
+    for (Limelight limelight : Limelight.values()) {
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/vision/SimulatedVision.java:110]: comment what this loop is doing

File: src/main/java/frc/robot/subsystems/vision/SimulatedVision.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+      writeToTable(
+          limelight.getSimulationCamera(limelight).getAllUnreadResults(),
+          limelight.getLimelightTable(limelight),
+          limelight);
+      inputs.limelightTargets[limelight.getId()] = getNumberOfAprilTags(limelight);
+    }
+    super.updateInputs(inputs);
+  }
+
+  private void writeToTable(
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/vision/SimulatedVision.java:120]: add a docstring

File: src/main/java/frc/robot/subsystems/vision/SimulatedVision.java (hunk 5 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/vision/SimulatedVision.java:61]: nah im just dumb

File: src/main/java/frc/robot/subsystems/vision/SimulatedVision.java (hunk 6 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/vision/SimulatedVision.java:69]: 👍

File: src/main/java/frc/robot/subsystems/swerve/odometryThread/OdometryThread.java (hunk 3 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/swerve/odometryThread/OdometryThread.java:51]: #82

File: src/main/java/frc/robot/subsystems/vision/SimulatedVision.java (hunk 8 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/vision/SimulatedVision.java:120]: ok

File: src/main/java/frc/robot/subsystems/vision/VisionInterface.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
+package frc.robot.subsystems.vision;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import frc.robot.extras.vision.MegatagPoseEstimate;
+import frc.robot.subsystems.vision.VisionConstants.Limelight;
+import org.littletonrobotics.junction.AutoLog;
+
+public interface VisionInterface {
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/vision/VisionInterface.java:8]: YOO GET SOME DOCSTRINGS IN THIS INTERFACE

File: src/main/java/frc/robot/subsystems/vision/VisionInterface.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    public boolean[] limelightSeesAprilTags = new boolean[Limelight.values().length];
+
+    public Pose2d[] limelightCalculatedPose = new Pose2d[Limelight.values().length];
+    public Pose2d limelightLastSeenPose = new Pose2d();
+    public double[] limelightAprilTagDistance = new double[Limelight.values().length];
+
+    public double[] limelightTimestamp = new double[Limelight.values().length];
+  }
+
+  default void updateInputs(VisionInputs inputs) {}
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/vision/VisionInterface.java:26]: I don't think we should use `default` in our interfaces. IMO it kind of goes against what interfaces should be, plus we shouldn't actually have any methods we don't implement.

File: src/main/java/frc/robot/subsystems/vision/Vision.java [hunk 11]
```diff
@@ -1,54 +1,60 @@
...
+
 package frc.robot.subsystems.vision;
 
 import edu.wpi.first.math.geometry.Pose2d;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.subsystems.vision.VisionConstants.Limelight;
 import org.littletonrobotics.junction.AutoLogOutput;
 import org.littletonrobotics.junction.Logger;
 
 public class Vision extends SubsystemBase {
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/vision/Vision.java:11]: Why did we move away from adding `Subsystem` after the names of our subsystems?

File: src/main/java/frc/robot/subsystems/vision/VisionConstants.java [hunk 12]
```diff
@@ -1,9 +1,66 @@
...
+import edu.wpi.first.apriltag.AprilTagFieldLayout;
+import edu.wpi.first.apriltag.AprilTagFields;
+import edu.wpi.first.math.geometry.Rotation3d;
+import edu.wpi.first.math.geometry.Transform3d;
+import edu.wpi.first.math.geometry.Translation3d;
 import edu.wpi.first.math.util.Units;
+import org.photonvision.PhotonCamera;
 
 public final class VisionConstants {
+  public enum Limelight {
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/vision/VisionConstants.java:12]: I like the enum. I was gonna tell you to change it to `Limelights` but I'm pretty sure this is correct convention lol.

File: src/main/java/frc/robot/extras/vision/FiducialObservation.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
+package frc.robot.extras.vision;
+
+import edu.wpi.first.util.struct.Struct;
+import edu.wpi.first.util.struct.StructSerializable;
+import java.nio.ByteBuffer;
+
+public class FiducialObservation implements StructSerializable {
```
Comment (JacksonElia) [inline src/main/java/frc/robot/extras/vision/FiducialObservation.java:7]: What is this for? Also needs to be commented

File: src/main/java/frc/robot/subsystems/vision/VisionInterface.java (hunk 10 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/vision/VisionInterface.java:26]: then open up a pr or an issue and change it

Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/vision/VisionInterface.java:26]: tyty

File: src/main/java/frc/robot/subsystems/vision/Vision.java (hunk 11 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/vision/Vision.java:11]: bc visionsubsystem is icky

**Pull request closed without merging**